RFC: Wakeup Controller (WUC) Driver

Albort12138 opened the pull request:
Fix: #100853

#100866 (c89a262d39a8933f34b22e40ddd6b2865d904075)demonstrates an implementation  of this new device driver.
#100865 (91fa5e4cc694666255320fb3d03714e4f2f8370c) demonstrates an integration with other device driver.

### Overview
This proposal introduces a new device driver for the Wakeup Controller (WUC). It involves:
- Adding new APIs to manage the WUC peripheral.

---

### Motivation
In some MCUs, waking up from low-power modes requires special configuration of a dedicated peripheral called the Wakeup Controller (WUC). Currently, Zephyr supports this functionality in two ways:

1. **Implemented in interrupt controllers** (e.g., [`wuc_ite_it51xxx.c`](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/interrupt_controller/wuc_ite_it51xxx.c)):
   This approach lacks a unified interface for WUC control. For example, if different devices share the same peripheral but use different WUC implementations, supporting a new device often requires modifying its driver (e.g., [`espi_it8xxx2.c`](https://github.com/zephyrproject-rtos/zephyr/blob/fd63bf93a5a18bf0dc0a18b67b821bfdd68ab5b9/drivers/espi/espi_it8xxx2.c#L2684C1-L2688C7)).

2. **Implemented in SoC-specific power management code**(e.g., [`power.c`](https://github.com/zephyrproject-rtos/zephyr/blob/fd63bf93a5a18bf0dc0a18b67b821bfdd68ab5b9/soc/nxp/mcx/mcxw/mcxw7xx/power.c#L25)):  
   This method only supports a limited set of peripherals and lacks flexibility. If an application needs to use an unsupported peripheral as a wakeup source, additional code changes are required, and developers must be familiar with the MCU internals.

Both approaches have common issues:
- WUC configuration is typically done during initialization, with no flexible API for user-space control.
- The relationship between peripherals and WUC is hidden in code rather than expressed in devicetree, reducing transparency and flexibility.

**Problems solved by this proposal:**
- Introduce a generic WUC device driver with APIs that can be reused by other drivers, reducing the need for driver-specific changes when supporting new SoCs.
- Represent the peripheral-to-WUC mapping in devicetree instead of hardcoding it in source code.

---

### Design Details
The proposal introduces a new Wakeup Controller (WUC) driver API and integrates it with Zephyr’s power management subsystem.

**[New WUC Driver API](https://github.com/nxp-upstream/zephyr/blob/create-wakeup-controller/include/zephyr/drivers/wuc.h):**
- `wuc_enable_wakeup_source()` – Enable a specific wakeup source.
- `wuc_disable_wakeup_source()` – Disable a specific wakeup source.
- `wuc_check_wakeup_triggered()` – Check if a wakeup was triggered by a specific source.
- `wuc_record_wakeup_source_triggered` - Recode a specific wakeup source occurs.
- `wuc_clear_wakeup_source_triggered` - Clear triggered wakeup source.

ZhaoxiangJin commented on include/zephyr/pm/device.h:
```diff
@@ -205,6 +215,10 @@ struct pm_device_base {
 	const struct device *domain;
 #endif /* CONFIG_PM_DEVICE_POWER_DOMAIN */
+#if defined(CONFIG_PM_WAKEUP_CONTROLLER) || defined(__DOXYGEN__)
+	uint8_t wuc_count;
```
What does `wuc_count `represent? The number of wakeup unit controllers on an SoC? Why is it placed in pm_device?

Albort12138 committed:
drivers: wuc: introduce Wakeup Controller (WUC) driver API

Introduce a new Wakeup Controller (WUC) driver API to manage wakeup
sources in low-power scenarios. The WUC API provides a standardized
interface for enabling, disabling, checking, recording and clearing
wakeup sources that can bring a device out of low-power states.

Signed-off-by: Albort Xue <yao.xue@nxp.com>

JarmouniA reviewed (commented):
This PR shall implement the API for at least one platform to demonstrate that it's usable.
Even better, you (or someone else) should try implementing it for a second platform to demonstrate that it's generic enough to be adopted by other vendors.

Edit: just saw #100865, it should be mentioned in the description, or even better added as a separate commit in this PR.

JarmouniA reviewed (commented):
A couple of remarks:
- You have to specify from what Zephyr state the system is waking up: Poweroff or low-power states, or both.
- you have to demonstrate that you have looked sufficiently at similar IPs of other vendors, so that the semantics of the core API can become sufficiently vendor-agnostic. Of course, there can be vendor-specific extensions to cover the features that are hard to generalize.

JarmouniA commented on include/zephyr/drivers/wuc.h:
```diff
@@ -0,0 +1,343 @@
+ * @retval 0 If successful
+ * @retval -errno Negative errno code on failure
+ */
+__syscall int wuc_enable_wakeup_source(const struct device *dev, uint32_t id);
```
no need for syscalls since the API is used by drivers.

Albort12138 replied:
Removed

JarmouniA commented on include/zephyr/drivers/wuc/wuc_nxp_llwu.h:
```diff
@@ -0,0 +1,119 @@
+/* PIN wakeup sources (PIN 0-16) */
+#define NXP_LLWU_PIN_0_RISING    NXP_LLWU_PIN_WAKEUP(1, NXP_LLWU_PIN_EDGE_RISING)
+#define NXP_LLWU_PIN_0_FALLING   NXP_LLWU_PIN_WAKEUP(1, NXP_LLWU_PIN_EDGE_FALLING)
+#define NXP_LLWU_PIN_0_ANY       NXP_LLWU_PIN_WAKEUP(1, NXP_LLWU_PIN_EDGE_ANY)
```
better done in DT as child nodes of WUC.
ex. https://github.com/zephyrproject-rtos/zephyr/blob/cdb7debfb726569440fc22ce6673de2b86c728f7/dts/arm/st/u5/stm32u5.dtsi#L950

Albort12138 commented on drivers/wuc/wuc_nxp_llwu.c:
```diff
@@ -0,0 +1,238 @@
+
+	K_SPINLOCK(&data->lock) {
+		for (int i = 0; i < 16; i++) {
+			if (IS_BIT_SET(data->enabled_pins, i) == true) {
```
Thanks your comments, updated

Albort12138 commented on drivers/wuc/wuc_nxp_llwu.c:
```diff
@@ -0,0 +1,238 @@
+
+static struct mcux_llwu_data llwu_data;
+
+static const struct wuc_driver_api mcux_llwu_api = {
```
Updated

Albort12138 commented:
Hi @bjarki-andreasen @rruuaanng @JarmouniA , The PR has been updated. Please take a look when you have time. Thank you very much

bjarki-andreasen commented on include/zephyr/drivers/wuc.h:
```diff
@@ -0,0 +1,331 @@
+typedef int (*wuc_api_record_wakeup_source_triggered)(const struct device *dev, uint32_t id);
+
+__subsystem struct wuc_driver_api {
+	wuc_api_enable_wakeup_source enable_wakeup_source;
```
APIs are already name spaced, so _wakeup_source is redundant, could just be enable, disable, triggered, clear etc.

Albort12138 replied:
Good suggestion, will update

bjarki-andreasen commented on include/zephyr/drivers/wuc.h:
```diff
@@ -0,0 +1,331 @@
+ * @retval 0 If successful.
+ * @retval -errno Negative errno code on failure.
+ */
+static inline int wuc_record_wakeup_source_triggered(const struct device *dev, uint32_t id)
```
I believe this means "trigger the wakeup source"? I don't immediately understand what "record" means in this context

Albort12138 replied:
In some WUC implementations, whether a given wakeup source actually woke the system is not reflected in the WUC peripheral itself. In those cases the wakeup source must be manually recorded into the WUC device driver so it can be queried later. For example, we have a WUC called "wuu" and an internal peripheral interrupt ( take timer interrupt as example) used as a wakeup source. When the LPTMR interrupt wakes the MCU, there is no status bit in wuu indicating the LPTMR caused the wakeup. I expect the record API to be used to update the wuu device driver so that subsequent checks will show the LPTMR wakeup was triggered.

bjarki-andreasen commented on include/zephyr/drivers/wuc.h:
```diff
@@ -0,0 +1,331 @@
+ * @retval 0 If successful.
+ * @retval -errno Negative errno code on failure.
+ */
+static inline int wuc_clear_wakeup_source_triggered(const struct device *dev, uint32_t id)
```
This API should be optional, some wakeup controllers can not check, clear or trigger the wakeup source, they are simply "triggered" if some hardware event is pending.

Albort12138 commented on include/zephyr/drivers/wuc.h:
```diff
@@ -0,0 +1,331 @@
+ * Example usage:
+ *
+ *	const struct wuc_dt_spec spec = WUC_DT_SPEC_GET_BY_IDX(DT_NODELABEL(n), 0);
+ *	// Initializes 'spec' to:
```
Fixed

Albort12138 committed:
tests: drivers: wuc: add WUC API test suite

Add a test suite for the Wakeup Controller (WUC) driver API that
validates the core functionality including enabling, disabling
wakeup sources.

The test suite includes:
- Test for enabling wakeup sources
- Test for disabling wakeup sources

Signed-off-by: Albort Xue <yao.xue@nxp.com>

Albort12138 committed:
drivers: wuc: add NXP LLWU (Low-Leakage Wakeup Unit) driver

Add a Wakeup Controller driver for NXP's LLWU
peripheral found on MCX-C series MCUs.

Signed-off-by: Albort Xue <yao.xue@nxp.com>

JarmouniA commented on dts/bindings/wuc/wuc-device.yaml:
```diff
@@ -0,0 +1,10 @@
+  wakeup-ctrls:
+    type: phandle-array
+    description: |
+      Phandle to the wakeup controller used by this device.
```
```suggestion
      Phandle of corresponding wakeup controller.
```

Albort12138 replied:
Updated

JarmouniA commented on tests/drivers/wuc/wuc_api/testcase.yaml:
```diff
@@ -0,0 +1,11 @@
+    - drivers
+    - wuc
+tests:
+  drivers.wuc.api.nxp:
```
```suggestion
  drivers.wuc.api
```

Albort12138 commented on dts/bindings/wuc/wuc-controller.yaml:
```diff
@@ -0,0 +1,12 @@
+# Copyright 2025 NXP
+# SPDX-License-Identifier: Apache-2.0
+
+description: Wakeup Controller
```
Updated

Albort12138 commented on dts/bindings/wuc/wuc-device.yaml:
```diff
@@ -0,0 +1,10 @@
+# Copyright 2025 NXP
+# SPDX-License-Identifier: Apache-2.0
+
+description: This file needs to be included by devices that need a wake up controller.
```
Updated

JarmouniA commented on include/zephyr/drivers/wuc.h:
```diff
@@ -0,0 +1,312 @@
+ * @retval -errno Negative errno code on failure.
+ * @retval -ENOSYS if the interface is not implemented.
+ */
+static inline int wuc_check_wakeup_source_triggered(const struct device *dev, uint32_t id)
```
I think so.

bjarki-andreasen commented on tests/drivers/wuc/wuc_api/testcase.yaml:
```diff
@@ -0,0 +1,10 @@
+    - wuc
+tests:
+  drivers.wuc.api:
+    depends_on: wuc
```
please filter on the existence of `DT_PATH(zephyr_user), wakeup_ctrls`

I believe this should work:
```
drivers.wuc.api:
  filter: dt_node_has_prop("/zephyr,user", "wakeup-ctrls")
```

Usually its enough to test just one board per soc, so its typical to only include an overlay for one of them.

Albort12138 replied:
Updated.

bjarki-andreasen reviewed (changes requested):
Looking really good!

ceolin commented on dts/arm/nxp/mcx/nxp_mcxc_common.dtsi:
```diff
@@ -389,6 +389,14 @@
 			voltage-reference = <2>;
 		};
+
+		llwu: llwu@0x4007c000 {
```
Does it works ? I thought you couldn't have `0x` in the address.

Albort12138 replied:
Fixed

ceolin reviewed (dismissed):
It looks good to me. Minor comments in the driver only.

Albort12138 committed:
doc: add WUC peripheral documentation

Add WUC peripheral docs and list the WUC section in the
peripherals index.

Signed-off-by: Albort Xue <yao.xue@nxp.com>

bjarki-andreasen reviewed (changes requested):
See comment

Albort12138 commented on drivers/wuc/wuc_nxp_llwu.c:
```diff
@@ -0,0 +1,193 @@
+					data->triggered_pin_sources |= BIT(i);
+					LLWU_ClearExternalWakeupPinFlag(config->base, i + 1);
+				} else {
+					data->triggered_pin_sources &= ~BIT(i);
```
Fixed

Albort12138 commented on dts/bindings/wuc/nxp,llwu.yaml:
```diff
@@ -0,0 +1,24 @@
+# Copyright 2025 NXP
```
All new created files updated to 2026.

Albort12138 commented:
Hi @bjarki-andreasen and @ceolin,
I would appreciate your help reviewing this PR. I’m hoping it can be merged in time for the 4.4 feature freeze.
Hi @sylvioalves and @henrikbrixandersen,
I’ve added a code snippet in #103863, and based on that, I believe the existing WUC APIs should already satisfy your use cases. Your review and comments would be greatly appreciated.

sylvioalves commented:
LGTM on driver API. Left a comment in Kconfig entry.

Albort12138 commented:
Hi @ceolin, @henrikbrixandersen, please take a look at this PR, thank you very much.

Albort12138 commented:
Hi @henrikbrixandersen, my understanding is @sylvioalves has reviewed the PR and confirmed that the existing APIs meet his requirements. For reference, please see #103863.
From your perspective, are there any additional concerns? If not, may we remove the “Architecture Review” label and proceed with this PR?
